### PR TITLE
Collijk/integration/regression cleanup

### PR DESCRIPTION
**Pull request by collijk**
Top-level summary:  This integrates the regression stage of the pipeline!

Important changes/fixes:
- Fixed some import bugs that snuck in with the last PR
- Updated regression specification to current production settings
- Removed regression diagnostics because they don't get used.
- Add classmethod constructors to the fit and regression specifications
- Write files by draw instead of by location draw
- Remove a bunch of layers of abstraction and delete a lot of not useful code.
- Lift saving and loading coeficients out of the modeling code (BetaRegressor)
- Add notebook for integration testing of the regression stage

Other things here:
- Tests for the fit stage (thanks Mike!)

Important Note: We don't get exact equality with the production pipeline.  The deviations are due to numerical roundoff and the complex operations going on under the hood in the regression.  The most important output here, beta hat, has errors that are on the order of 1e-4 and are normally distributed around zero, which I think is totally acceptable.

### CODE REVIEW

File: src/covid_model_seiir_pipeline/cli.py [hunk 1]
```diff
@@ -2,11 +2,18 @@
...
 from loguru import logger
 
-from covid_model_seiir_pipeline import regression, ode_fit, forecasting, utilities
+from covid_model_seiir_pipeline import utilities
+from covid_model_seiir_pipeline.ode_fit.specification import FitSpecification
+from covid_model_seiir_pipeline.ode_fit.main import do_beta_fit
+from covid_model_seiir_pipeline.regression.specification import RegressionSpecification
+from covid_model_seiir_pipeline.regression.main import do_beta_regression
+from covid_model_seiir_pipeline.forecasting.specification import ForecastSpecification
+from covid_model_seiir_pipeline.forecasting.main import do_beta_forecast
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/cli.py:11]: Whoops, this was an oversight on my part missing these imports. Apologies

File: src/covid_model_seiir_pipeline/cli.py [hunk 2]
```diff
@@ -62,7 +62,6 @@ def fit(run_metadata,
     fit_spec.data.location_set_version_id = locations_set_version_id
     fit_spec.data.location_set_file = location_set_file
     fit_spec.data.output_root = str(run_directory)
-    fit_spec.dump(run_directory / 'fit_specification.yaml')
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/cli.py:58]: why was this removed? was the line redundant with another one?

File: src/covid_model_seiir_pipeline/cli.py [hunk 3]
```diff
@@ -125,7 +123,6 @@ def regress(run_metadata,
     regression_spec.data.ode_fit_version = str(ode_fit_root)
     regression_spec.data.covariate_version = str(covariates_root)
     regression_spec.data.output_root = str(run_directory)
-    regression_spec.dump(run_directory / 'regression_specification.yaml')
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/cli.py:121]: same question as `fit.spec.dump()`

File: src/covid_model_seiir_pipeline/regression/workflow.py [hunk 4]
```diff
@@ -81,20 +79,17 @@ def get_task(self, regression_version: str):
 
 class RegressionWorkflow:
 
-    def __init__(self, regression_specification: RegressionSpecification,
-                 ode_fit_specification: FitSpecification):
-        self.regression_specification = regression_specification
-        self.ode_fit_specification = ode_fit_specification
-
+    def __init__(self, regression_version):
+        self.regression_version = regression_version
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/regression/workflow.py:83]: This must be a YYYY-MM-DD.VV string right? Or does it include the path leading up to it?

File: src/covid_model_seiir_pipeline/regression/data.py [hunk 5]
```diff
@@ -84,13 +85,12 @@ def load_covariate(self, covariate: str, location_ids: List[int]) -> pd.DataFram
...
         return covariate_df.loc[:, index_columns + [covariate]].set_index(index_columns)
 
-
     def load_covariates(self, covariates: Iterable[str], location_ids: List[int]) -> pd.DataFrame:
         covariate_data = []
         for covariate in covariates:
             if covariate != 'intercept':
                 covariate_data.append(self.load_covariate(covariate, location_ids))
-        covariate_data = pd.concat(covariate_data, axis=1)
+        covariate_data = reduce(lambda x, y: x.merge(y, left_index=True, right_index=True), covariate_data)
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/regression/data.py:94]: :partying_face:

File: src/covid_model_seiir_pipeline/regression/tasks/beta_regression.py [hunk 6]
```diff
@@ -118,7 +102,7 @@ def run_beta_regression(draw_id: int, regression_version: str) -> None:
     # -------------- BETA REGRESSION WITH LOADED COVARIATES -------------------- #
     # Convert inputs for beta regression
     mr = ModelRunner()
-
+    import pdb; pdb.set_trace()
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/regression/tasks/beta_regression.py]: TODO: ensure not in final PR

**Comment by miker985** [comment]
Soo many changes in this. Some bug fixes, others appear to be code motion and refactoring. It's hard to keep track of everything going on.

Going commit by commit helped, but I still don't feel confident I could describe all the things happening here.

File: src/covid_model_seiir_pipeline/cli.py (hunk 2 above)
Comment (collijk) [reply src/covid_model_seiir_pipeline/cli.py:58]: Consistency with other applications means that the main function should be responsible for everything besides the top level directory and the metadata.yaml file that lives within it.  Specifications are dumped in the main function now.

File: src/covid_model_seiir_pipeline/regression/data.py (hunk 5 above)
Comment (collijk) [reply src/covid_model_seiir_pipeline/regression/data.py:94]: I was sad to resort to a merge, but pd.concat didn't like the required broadcasting/dealing with nulls.

File: src/covid_model_seiir_pipeline/paths.py [hunk 7]
```diff
@@ -8,6 +8,9 @@
 from covid_shared.shell_tools import mkdir
 from loguru import logger
 
+DRAW_FILE_TEMPLATE = 'draw_{draw_id}.csv'
+LOCATION_FILE_TEMPLATE = 'location_{location_id}.csv'
```
Comment (megan-h-moore) [inline src/covid_model_seiir_pipeline/paths.py:12]: I might be missing something but I am not seeing where this is getting used.

File: src/covid_model_seiir_pipeline/regression/workflow.py [hunk 8]
```diff
@@ -53,7 +50,6 @@ def __init__(self, task_registry: Dict[str, BashTask]):
         )
         self.params = ExecutorParameters(
             max_runtime_seconds=1000,
-            j_resource=False,
```
Comment (megan-h-moore) [inline src/covid_model_seiir_pipeline/regression/workflow.py:56]: default is false so just want to make sure J access isn't actually needed.

File: src/covid_model_seiir_pipeline/paths.py (hunk 7 above)
Comment (collijk) [reply src/covid_model_seiir_pipeline/paths.py:12]: the location one isn't yet.

**Comment by collijk** [comment]
This is the distribution of relative error in beta hat over all time an locations for a single draw.  It's symmetric and small, and almost certainly due to the propagation of small roundoff errors in the regression.

image

**Comment by collijk** [comment]
image

This is a comparison of the coefficients from an identical run in the production pipeline.  The only significant deviations occur in the non-time-varying covariates whose values are small compared to the intercept in both situations (indicating they have almost no explanatory power).

File: specification_templates/regression.yaml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 data:
-  covariate_version: 'best'
-  ode_fit_version: 'best'
+  covariate_version: '2020_06_30.01'
+  ode_fit_version: 'latest'
```
Comment (miker985) [inline specification_templates/regression.yaml:3]: really, latest?

File: src/covid_model_seiir_pipeline/ode_fit/data.py [hunk 10]
```diff
@@ -16,6 +17,15 @@ def __init__(self, ode_paths: paths.ODEPaths,
         self.ode_paths = ode_paths
         self.infection_paths = infection_paths
 
+    @classmethod
+    def from_specification(cls, specification: FitSpecification) -> 'ODEDataInterface':
+        ode_paths = paths.ODEPaths(Path(specification.data.output_root), read_only=False)
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/ode_fit/data.py:22]: why is this mutable (not read_only)?

File: src/covid_model_seiir_pipeline/regression/task.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
+from argparse import ArgumentParser, Namespace
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/regression/task.py:1]: was this just a rename of `beta_regression.py`?

Comment (collijk) [reply src/covid_model_seiir_pipeline/regression/task.py:1]: Sorry, this was super sloppy of me.

I axed the diagnostics task, at which point we didn't need a subpackage for tasks.  This is substantially rewritten from the original though.
